eSPI support implementation

LKomaryanskiy opened the pull request:
Implemented eSPI support, which was introduced with GICv3.1. This is an essential feature for Gen5, as it utilizes this type of interrupt for critical devices, such as HSCIF, which is required for booting Xen.

LKomaryanskiy committed:
xen/arm/irq: add handling for IRQs in the eSPI range

Currently, Xen does not support eSPI interrupts, leading
to a data abort when such interrupts are defined in the DTS.

This patch introduces a separate array to initialize up to
1024 interrupt descriptors in the eSPI range and adds the
necessary defines and helper function. These changes lay the
groundwork for future implementation of full eSPI interrupt
support. As this GICv3.1 feature is not required by all vendors,
all changes are guarded by ifdefs, depending on the corresponding
Kconfig option.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

LKomaryanskiy committed:
xen/arm/irq: allow assignment/releasing of eSPI interrupts

The current checks don't allow us to assign or release interrupts
with INTID greater than 1024. This patch adds an additional condition
to check whether the IRQ number is in the eSPI range and allows it to
be assigned to Xen and domains if it is.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

LKomaryanskiy committed:
xen/arm: gic/irq: permit routing of eSPI interrupts to Xen and domains

Current checks prevent routing of interrupts from the eSPI range to Xen
or guest domains. The addition of the is_espi condition allows verification
of whether a given index falls within the eSPI range, enabling routing of
such interrupts.

Existing behavior remains unchanged for configurations where eSPI support
is disabled, as in this case, is_espi always returns false.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

LKomaryanskiy committed:
xen/arm/irq: allow eSPI processing in the do_IRQ function

The do_IRQ() function is the main handler for processing IRQs.
Currently, due to restrictive checks, it does not process interrupt
numbers greater than 1024. This patch updates the condition to allow
the handling of interrupts from the eSPI range.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

LKomaryanskiy committed:
xen/arm: gicv3: modify ICH_LR_PHYSICAL_MASK to allow eSPI processing

To properly deactivate guest interrupts and allow them to be retriggered
after the initial trigger, the LR needs to be updated. The current
implementation ignores interrupts outside the range specified by the mask
0x3FF, which only covers IRQ numbers up to 1023. To enable processing of
eSPI interrupts, this patch updates the mask to 0x13FF.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

LKomaryanskiy committed:
xen/arm: gicv3: refactor obtaining GIC addresses for common operations

Currently, many common functions perform the same operations to calculate
GIC register addresses. This patch consolidates the similar code into
a separate helper function to improve maintainability and reduce duplication.
This refactoring also simplifies the implementation of eSPI support in future
changes.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

LKomaryanskiy committed:
xen/arm: gicv3: implement handling of GICv3.1 eSPI

Introduced appropriate register definitions, helper macros,
and initialization of required GICv3.1 distributor registers
to support eSPI. This type of interrupt is handled in the
same way as regular SPI interrupts, with the following
differences:

1) eSPIs can have up to 1024 interrupts, starting from the
beginning of the range, whereas regular SPIs use INTIDs from
32 to 1019, totaling 988 interrupts;
2) eSPIs start at INTID 4096, necessitating additional interrupt
index conversion during register operations.

In case if appropriate config is disabled, or GIC HW doesn't
support eSPI, the existing functionality will remain the same.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

LKomaryanskiy committed:
xen/arm: vgic: add resource management for extended SPIs

This change introduces resource management in the VGIC to handle
extended SPIs introduced in GICv3.1. The pending_irqs and
allocated_irqs arrays are resized to support the required
number of eSPIs, based on what is supported by the hardware and
requested by the guest. A new field, ext_shared_irqs, is added
to the VGIC structure to store information about eSPIs, similar
to how shared_irqs is used for regular SPIs.

Since the eSPI range starts at INTID 4096 and INTIDs between 1025
and 4095 are reserved, helper macros are introduced to simplify the
transformation of indices and to enable easier access to eSPI-specific
resources. These changes prepare the VGIC for processing eSPIs as
required by future functionality.

The initialization and deinitialization paths for vgic have been updated
to allocate and free these resources appropriately. Additionally,
updated handling of INTIDs greater than 1024, passed from the toolstack
during domain creation, and verification logic ensures only valid SPI or
eSPI INTIDs are used.

The existing SPI behavior remains unaffected when guests do not request
eSPIs, GIC hardware does not support them, or the CONFIG_GICV3_ESPI
option is disabled.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

LKomaryanskiy committed:
xen/arm: domain_build: adjust Dom0 IRQ handling to support eSPIs

The Dom0 configuration logic in create_dom0() has been updated
to account for extended SPIs when supported by the hardware and
enabled with CONFIG_GICV3_ESPI. These changes ensure the proper
calculation of the maximum number of SPIs and eSPIs available for Dom0.

When eSPIs are supported by the hardware and CONFIG_GICV3_ESPI is
enabled, the maximum number of eSPI interrupts is calculated using
the ESPI_BASE_INTID offset (4096) and limited at 1024, with 32 IRQs
subtracted. To ensure compatibility with non-Dom0 domains, this
adjustment is applied by the toolstack during domain creation, while
for Dom0 it is handled directly during VGIC initialization. If eSPIs
are not supported, the calculation defaults to using the standard SPI
range, with a maximum value of 992 interrupt lines as it works now.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

LKomaryanskiy committed:
xen/arm: vgic-v3: add emulation of GICv3.1 eSPI registers

Implemented support for GICv3.1 extended SPI registers for vGICv3,
allowing the emulation of eSPI-specific behavior for guest domains.
The implementation includes read and write emulation for eSPI-related
registers (e.g., GICD_ISENABLERnE, GICD_IROUTERnE, and others),
following a similar approach to the handling of regular SPIs.

The eSPI registers, previously located in reserved address ranges,
are now adjusted to support MMIO read and write operations correctly
when CONFIG_GICV3_ESPI is enabled.

The availability of eSPIs and the number of emulated extended SPIs
for guest domains is reported by setting the appropriate bits in the
GICD_TYPER register, based on the number of eSPIs requested by the
domain and supported by the hardware. In cases where the configuration
option is disabled, the hardware does not support eSPIs, or the domain
does not request such interrupts, the functionality remains unchanged.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity commented on xen/arch/arm/Kconfig:
```diff
@@ -106,6 +106,16 @@ config GICV3
+	bool "Extended SPI range support"
+	depends on GICV3 && !NEW_VGIC
+	default n
+	---help---
```
Should the `help` property be written like that? I've never seen such way before

LKomaryanskiy replied:
Some configs even in this file contain "help" section, described like this. But yes, you are right, it is better to define it more typically. So fixed, thank you:)

firscity commented on xen/arch/arm/include/asm/irq.h:
```diff
@@ -45,15 +45,31 @@ struct arch_irq_desc {
+#ifdef CONFIG_GICV3_ESPI
+#define ESPI_BASE_INTID 4096
+#define ESPI_MAX_INTID  5119
+#define ESPI_INTID2IDX(intid) ((intid) - ESPI_BASE_INTID)
```
nit: I think blank line will look good here, they has different meaning (constants and macros) and has different tab indent

LKomaryanskiy replied:
Added dividing whitespace between constants and macros. Now, it looks better, thank you.

firscity commented on xen/arch/arm/include/asm/irq.h:
```diff
@@ -69,6 +85,15 @@ static inline bool is_lpi(unsigned int irq)
     return irq >= LPI_OFFSET;
 }
 
+static inline bool is_espi(unsigned int irq)
```
As I see, this function is used only in code, that is under `CONFIG_GICV3_ESPI`, maybe it is better to move it under this config completely?

firscity commented on xen/arch/arm/irq.c:
```diff
@@ -604,8 +604,8 @@ int setup_irq(unsigned int irq, unsigned int irqflags, struct irqaction *new)
-    /* For now, we can only route SPIs to the guest */
-    return (irq >= NR_LOCAL_IRQS) && (irq < gic_number_lines());
+    /* For now, we can only route SPIs and eSPIs to the guest */
+    return (((irq >= NR_LOCAL_IRQS) && (irq < gic_number_lines())) || is_espi(irq));
```
Oh, I see where `is_espi()` is also used

Also reminder about `modify check in release_guest_irq` in commit message

LKomaryanskiy replied:
`Also reminder about modify check in release_guest_irq in commit message`

Oh, I'm sorry, I mistakenly leftover it in commit while squashing changes. Removed, thank you!

firscity commented on xen/arch/arm/gic.c:
```diff
@@ -111,7 +111,7 @@ static void gic_set_irq_priority(struct irq_desc *desc, unsigned int priority)
 {
     ASSERT(priority <= 0xff);     /* Only 8 bits of priority */
-    ASSERT(desc->irq < gic_number_lines());/* Can't route interrupts that don't exist */
+    ASSERT(desc->irq < gic_number_lines() || is_espi(desc->irq));/* Can't route interrupts that don't exist */
```
Commit message:
I think it is better to expand it (to add more context, little duplication is not a problem) and also not to use starting phrase "This patch". AFAIK it is not a good approach in describing changes, IIRC it is also mentioned in "Describe your changes" for Linux Kernel.

LKomaryanskiy replied:
Updated commit message with additional details and without "this patch" construction"

firscity commented on xen/arch/arm/gic.c:
```diff
@@ -375,7 +375,7 @@ void gic_interrupt(struct cpu_user_regs *regs, int is_fiq)
         irq = gic_hw_ops->read_irq();
 
-        if ( likely(irq >= 16 && irq < 1020) )
+        if ( likely((irq >= 16 && irq < 1020) || is_espi(irq)) )
```
Maybe `is_espi()` should be outside of `likely()` block, since it may be called even if ESPI config is turned off. I believe, that compiler will optimize it, since it returns false in such case unconditionally, but anyway...

LKomaryanskiy replied:
Yes, I also was not sure if it is really needed in likely, moreover, taking into account whether config will be disabled. So, moved from likely.

firscity commented on xen/arch/arm/include/asm/vgic.h:
```diff
@@ -148,8 +148,12 @@ struct vgic_dist {
     spinlock_t lock;
     uint32_t ctlr;
     int nr_spis; /* Number of SPIs */
+    int nr_espis; /* Number of extended SPIs */
```
Maybe this member should also be under `#ifdef`? It will require additional changes to handling, but will guarantee that previous behavior remains as it was.

LKomaryanskiy replied:
Initially, I didn't put this member in ifdefs, to reduce the number of total ifdefs, because it requires guarding all places where this member is used. But it looks like now, it is not a big problem, so could you please check a new version? Is this okay to have such number of ifdefs? I was afraid that it may affect code readability.  But now, both options are okay for me :)

firscity reviewed (commented):
Incredible job, Leonya!
No critical comments from my side, only cosmetic or small enhancements, I'm impressed!

LKomaryanskiy commented:
@firscity thank you for your close review and for your time!
I updated the PR with the following fixes:
- Removed mistakenly added part of commit message;
- Updated commit message for routing IRQs with additional details;
- Changed 'help' section in Kconfig to be more convenient and regular;
- Other minor fixes (codestyle, moving from likely construction, etc)

LKomaryanskiy commented:
Sorry, I left the redundant text in one commit message. In the last version, I just removed it

firscity reviewed (approved):
Good, thank you. Please, have my:
`Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>`

I think it is ready to be in our branch.

dimaprkp4k commented on xen/arch/arm/domain_build.c:
```diff
@@ -2121,13 +2121,28 @@ void __init create_dom0(void)
+    if ( gic_number_espis() > 0 )
+        dom0_cfg.arch.nr_spis = ESPI_BASE_INTID + min(gic_number_espis(), 1024U) - 32;
+    else
+        dom0_cfg.arch.nr_spis = min(gic_number_lines(), (unsigned int) 992) - 32;
```
I would like to replace "(unsigned int) 992" with "992U".

LKomaryanskiy replied:
Sure, I'll replace it. Basically, this code is a copy-paste from the original code that maintains regular SPIs, but I'll fix it anyway. Thank you for your review.

LKomaryanskiy commented:
@firscity and @dimaprkp4k thank you for your close review!
Updates in the last version:
- Added RB tags from @firscity 
- Minor code-style fix: Replaced  `(unsigned int) 992` with `992U` to ensure code compliance.
